src/quo2/foundations/colors.cljs

```diff
@@ -260,7 +257,34 @@
   [s]
   (and (string? s) (string/starts-with? s "#")))
 
-(def custom-color
+(def resolve-color
+  "(resolve-color color theme opacity)
+   color   hex string or keyword (resolves from custom, network and semantic colors)
+   theme  :light/:dark
+   opacity 0-100 (optional) - if set theme is ignored and goes to 50 suffix internally"
+  (memoize
```

Comment:
sure I can add this. it should be like this?
```
(defn resolve-color
     ([color theme]
      (resolve-color color theme nil))
     ([color theme opacity]
      (let [suffix           (cond opacity         50
                                   (= :dark theme) 60
                                   :else           50)
            resolve-keyword? (keyword? color)
            resolved-color   (if resolve-keyword?
                               (get-from-colors-map color suffix)
                               color)]
        (if opacity
          (alpha resolved-color (/ opacity 100))
          resolved-color))))
```

`(def memo-resolve-color (memoize resolve-color))`

Reply:
btw, we should have some guidance of when to use each. or the non memo is just for testing?

Reply:
@J-Son89, similar to what we do with the `icon` component, or the `i18n/label` function, it's the opposite of the code you shared previously.

The final implementation would be:

- public var `resolve-color`
- private var `do-resolve-color`, or `non-memo-resolve-color`, or `resolve-color-1` (quite common in the lisp tradition), etc.

Having the non-memoized function defined separately from the public & memoized one gives the developer the chance to change implementations in the REPL more conveniently. It's not a big deal, but I think it's a sensible thing to do, a good practice.

For instance:

1. `resolve-color` depends on `colors-map`, and `colors-map` depends on `customization` and `networks`.
2. If in a REPL session I call `(resolve-color :blue :light)`, the value will be cached in memory, and if I call with the same args again, the same value will be returned.
3. If I go ahead and change what `:blue` means in the `colors-map`, evaluate in the REPL the new `colors-map` var, the new blue value won't be used because `resolve-color` will return the cached value. For the new value to be returned, I need now to also re-evaluate the var `resolve-color` to reset the memoization cache. This makes iterating in the implementation less convenient, and sometimes frustrating because the dev thinks the implementation has a bug, when in fact it was just using cached results. This flow I'm sharing is happening outside the write file + rerun test, or the hot-reload, here I'm just using the REPL without even saving the file.

I hope what I said makes sense :)

Reply:
The other reason could be for testing as you said, although I think it's fine for `resolve-color`. One could argue that it's better to test the pure function only, and ignore the stateful/memoized one, since testing with memoization is just re-testing the core Clojure function behavior, which could be seen as pointless.

Reply:
@J-Son89 I agree about add a separate `def` for the memoized version. In production code we will always use the memoized one and just for testing the one that is not memoized.

I'd like to suggest a refactoring:
```clojure
(defn- resolve-color*
  ([color theme]
   (resolve-color* color theme nil))
  ([color theme opacity]
   (let [suffix (if (or opacity (= theme :light))
                  50
                  60)]
     (cond-> color
       (keyword? color) (get-from-colors-map suffix)
       opacity          (alpha (/ opacity 100))))))

(defn resolve-color (memoize resolve-color*))
```

The problem is I feel a little uncomfortable with the `suffix` var since we are calculating it even if we might not use it (it'll be unused if `(keyword? color)` is false :thinking:).  Instead, if we depend on the value of `suffix`:

```clojure
(defn- resolve-color*
  ([color theme]
   (resolve-color* color theme nil))
  ([color theme opacity]
   (let [suffix (cond
                  (not (keyword? color))        nil
                  (or opacity (= theme :light)) 50
                  :else                         60)]
     (cond-> color
       suffix  (get-from-colors-map suffix)
       opacity (alpha (/ opacity 100))))))
```

I think this one is not as clear as the previous one, as always, there are trade-offs.

IDK, just a suggestion, you could pick whatever you like more or just keep the current implementation.

Reply:
ok I like this idea, neat! 🧠